protocol/proto/encoder.go

```diff
@@ -44,14 +46,18 @@ func newEncoder(w io.Writer, opts BaseOptions) *encoder {
 	}
 	pool := opts.BytesPool()
 	return &encoder{
-		w:          w,
-		sizeBuffer: getByteSliceWithLength(sizeBufferSize, pool),
-		bytesPool:  pool,
+		w:              w,
+		sizeBuffer:     getByteSliceWithLength(sizeBufferSize, pool),
+		bytesPool:      pool,
+		maxMessageSize: opts.MaxMessageSize(),
 	}
 }
 
 func (e *encoder) Encode(m Marshaler) error {
 	size := m.Size()
+	if size > e.maxMessageSize {
+		return fmt.Errorf("message size %d is larger than maximum supported size %d", size, e.maxMessageSize)
```

Comment:
Food for thought: it might make sense to distinguish different types of errors so that callers can selectively choose to reset the encoder/decoder as needed.

Reply:
I could return a different error here and handle it differently in the writer, but I can just drop the data, and the user will have no way to know why it's being dropped and might retry again. I think it would be better to have the Buffer to reject the oversized data and return an error with the reason right in the beginning and save all the future work like trying to send it to all the consumers and then being dropped after the error in encoder, thoughts?

Reply:
Yeah sg